### PR TITLE
Containerized servings.py

### PR DESCRIPTION
**Pull request by l4b4r4b4b4**
Added Dockerfile & docker-compose.yml for containerized deployment.

Had to make a slight adjustment to `servings.py` changing host ip from `127.0.0.1` to `0.0.0.0`.

Hope this helps others with quick and convenient deployment and testing.

### CODE REVIEW

**Comment by vatsalaggarwal** [comment]
Thanks for this, we'll have a look in a bit... 

noticed that you were trying to add references... quick note on them:
i) we don't support cross-lingual cloning yet, so I think the german reference is unlikely to work
ii) we don't support references shorter than 30 seconds, and stuff is unclean/has background noise etc is very unlikely to work!

**Comment by l4b4r4b4b4** [comment]
> Thanks for this, we'll have a look in a bit...
> 
> noticed that you were trying to add references... quick note on them: i) we don't support cross-lingual cloning yet, so I think the german reference is unlikely to work ii) we don't support references shorter than 30 seconds, and stuff is unclean/has background noise etc is very unlikely to work!

The references were just for testing, what comes out, if I through in a German reference. Outcome: Funny :wink: 

Looking forward to actual fine-tuning capabilities. Are those LoRAs?

**Comment by vatsalaggarwal** [comment]
> The references were just for testing, what comes out, if I through in a German reference. Outcome: Funny 😉
lol expected

> Looking forward to actual fine-tuning capabilities. Are those LoRAs?
I am not sure if finetuning will be able to make that 20s sample with background noise work :P ... have you tried others that were longer than 30s and were clean that you're hoping to finetune on but couldn't get zero-shot to work? 

LoRAs - not sure yet... we're open to folks adding that, and I can provide some guidance!

**Comment by l4b4r4b4b4** [comment]
IF it was transformers I would be able to help quickly with LoRA PEFT. But as I see you have your very custom GPT implementation.

**Comment by vatsalaggarwal** [comment]
I am happy to help with those rough edges!

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+# install ffmpeg
+RUN wget https://johnvansickle.com/ffmpeg/builds/ffmpeg-git-amd64-static.tar.xz &&\
+    wget https://johnvansickle.com/ffmpeg/builds/ffmpeg-git-amd64-static.tar.xz.md5 &&\
+    md5sum -c ffmpeg-git-amd64-static.tar.xz.md5 &&\
+    tar xvf ffmpeg-git-amd64-static.tar.xz &&\
+    mv ffmpeg-git-*-static/ffprobe ffmpeg-git-*-static/ffmpeg /usr/local/bin/ &&\
+    rm -rf ffmpeg-git-*
+
+RUN pip install torch packaging wheel
```
Comment (sidroopdaska) [inline Dockerfile:19]: why install torch separately?

File: README.md [hunk 2]
```diff
@@ -11,6 +11,12 @@ We’re releasing MetaVoice-1B under the Apache 2.0 license, *it can be used wit
 
 Try out [demo](https://ttsdemo.themetavoice.xyz/)!
 
+## Quickstart - TLDR
+
+```bash
+docker-compose up -d && docker-compose ps && docker-compose logs -f
+```
+Explore exposed [API](http://localhost:8869/docs)!
```
Comment (sidroopdaska) [inline README.md:19]: nit: new line

File: .gitignore [hunk 3]
```diff
@@ -1,5 +1,6 @@
+.vscode
 *.pkl
-*.flac
+# *.flac
```
Comment (sidroopdaska) [inline .gitignore:3]: can we please remove these additional audio files under `assets/`

File: fam/llm/serving.py [hunk 4]
```diff
@@ -82,6 +82,7 @@ class TTSRequest:
 @app.post("/tts", response_class=Response)
 async def text_to_speech(req: Request):
     audiodata = await req.body()
+    print("audiodata := ", audiodata)
```
Comment (sidroopdaska) [inline fam/llm/serving.py:85]: remove?

File: fam/llm/serving.py [hunk 5]
```diff
@@ -92,6 +93,7 @@ async def text_to_speech(req: Request):
         tts_req = TTSRequest(**payload)
         with tempfile.NamedTemporaryFile(suffix=".wav") as wav_tmp:
             if tts_req.speaker_ref_path is None:
+                print("wav_tmp := ", wav_tmp)
```
Comment (sidroopdaska) [inline fam/llm/serving.py:96]: remove?

File: docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+          devices:
+            - driver: nvidia
+              count: 1
+              capabilities: [gpu]
+    runtime: nvidia
+    ipc: host
+    tty: true # enable colorized logs
+    build:
+      context: .
+    command: ['--huggingface_repo_id=metavoiceio/metavoice-1B-v0.1', "--port=80"]
```
Comment (sidroopdaska) [inline docker-compose.yml:29]: why port 80? you are later mapping it to 8869?
also, when i run this yaml, the server logging is lost.

**Review by sidroopdaska — CHANGES_REQUESTED** [review]
really appreciate you putting this together. a few changes requested

File: Dockerfile (hunk 1 above)
Comment (l4b4r4b4b4) [reply Dockerfile:19]: because it has to be installed before the other installs for it it work.

File: docker-compose.yml (hunk 6 above)
Comment (l4b4r4b4b4) [reply docker-compose.yml:29]: thats up to the user to decide...

**Comment by coder543** [comment]
I do not understand why this API is using a very non-standard "X-Payload" header to contain the request body, and despite the link to `/docs`, there is no real documentation there. The FastAPI implementation does not explicitly say that `TTSRequest` is the expected request type, likely because pulling it out of a header isn't a common approach, so the generated docs assume there are no parameters at all.

For anyone else trying to figure this out, here is a `curl` command that serves as a basic example with the way the API is currently structured:

```bash
curl -X POST http://localhost:8869/tts \
     -H "X-Payload: {\"text\": \"Hello, this is a test!\", \"guidance\": 3.0, \"top_p\": 0.95, \"speaker_ref_path\": \"assets/av
a.flac\"}" \
     -o output.mp3
```

**Comment by l4b4r4b4b4** [comment]
> I do not understand why this API is using a very non-standard "X-Payload" header to contain the request body, and despite the link to `/docs`, there is no real documentation there. The FastAPI implementation does not explicitly say that `TTSRequest` is the expected request type, likely because pulling it out of a header isn't a common approach, so the generated docs assume there are no parameters at all.
> 
> For anyone else trying to figure this out, here is a `curl` command that serves as a basic example with the way the API is currently structured:
> 
> ```shell
> curl -X POST http://localhost:8869/tts \
>      -H "X-Payload: {\"text\": \"Hello, this is a test!\", \"guidance\": 3.0, \"top_p\": 0.95, \"speaker_ref_path\": \"assets/av
> a.flac\"}" \
>      -o output.mp3
> ```
> 
> EDIT:
> 
> I have also discovered this:
> 
> ```
>     # NOTE: supports max. 220 characters atm.
>     # Long form synthesis coming soon...
>     MAX_CHARS = 220
> ```
> 
> so... that's fun. It's currently hardcoded to only support up to 220 characters of input. I can't see where the 220 character limit is actually being enforced, but I can't think of any use case where 220 characters is reliably large enough to use.
> 
> I am excited to see more options in the open source TTS space, so hopefully that limitation is lifted soon.

Yeah was wondering about X-Payload as well and thought of changing it to but just kept it as is in the end.

For the length, I think the current implementation expects you to basically feed the API a text in chunks, with a max char length of 220...

**Comment by vatsalaggarwal** [comment]
hey folks! we are grateful for your feedback, and contributions.

We are balancing fundamental improvements to the models, and improving the experience with the released model (such as optimizations, lower VRAM, Mac, releasing long-form, and releasing finetuning). We're trying to get them out as quickly as possible, but we are still a small team. That being said, we are actively hiring (https://metavoice.notion.site/Join-MetaVoice-e4c907cb6a2f4c33af2b148f635adda4?pvs=4). Alternatively, contributions are always welcome.

Thanks again you outlining the problems you mentioned, I assure you we're working hard to resolve them.

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    apt-get autoremove -y && \  
+    apt-get clean
+
+# install ffmpeg
+RUN wget https://johnvansickle.com/ffmpeg/builds/ffmpeg-git-amd64-static.tar.xz &&\
+    wget https://johnvansickle.com/ffmpeg/builds/ffmpeg-git-amd64-static.tar.xz.md5 &&\
+    md5sum -c ffmpeg-git-amd64-static.tar.xz.md5 &&\
+    tar xvf ffmpeg-git-amd64-static.tar.xz &&\
+    mv ffmpeg-git-*-static/ffprobe ffmpeg-git-*-static/ffmpeg /usr/local/bin/ &&\
+    rm -rf ffmpeg-git-*
```
Comment (djmaze) [inline Dockerfile:17]: In my Dockerfile, I used ffmpeg from the existing Ubuntu package sources and everything seemed to work well. Any reason this custom version should be installed?

This goes more to the maintainers I guess.

Comment (vatsalaggarwal) [reply Dockerfile:17]: Yeah, on our prod server, we were running ffmpeg that came with Amazon Linux’s Deep Learning AMI. It turned out that ffmpeg had a bug and was cutting audio short during mp3->wav conversions, so I added this just to be safe given prior scars :)

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+WORKDIR /app
+COPY requirements.txt requirements.txt
+# install python packages
+RUN pip install -r requirements.txt
+COPY . .
+RUN pip install -e .
+RUN apt-get remove -y build-essential && \  
+    apt-get autoremove -y && \  
+    apt-get clean  && \  
+    rm -rf /var/lib/apt/lists/*
```
Comment (djmaze) [inline Dockerfile:29]: Nitpicking, but removing this package in a separate docker layer is not going to help. It just adds an additional layer to the image and takes up even more space. Same with `pip cache purge` below.

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,33 @@
+FROM nvidia/cuda:12.1.0-devel-ubuntu22.04 as base
+RUN apt-get update && \  
+    apt-get upgrade -y 
+
+RUN apt-get update && \  
+    # apt-get upgrade -y && \
+    apt-get install -y python3.10 python3-pip git wget && \
```
Comment (djmaze) [inline Dockerfile:7]: `curl` is also needed (in `fam/llm/sample.py`).

File: docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+      - HF_HOME=/.hf-cache
+    logging:
+      options:
+        max-size: "100m"
+        max-file: "10"
+    healthcheck:
+      test: ["CMD", "curl", "-f", "http://metavoice-server:80/"]
+      interval: 1m
+      timeout: 20m
+      retries: 20
```
Comment (djmaze) [inline docker-compose.yml:45]: Could you also add a ui service as shown here? https://github.com/metavoiceio/metavoice-src/compare/main...groovybits:metavoice-src:docker#diff-e45e45baeda1c1e73482975a664062aa56f20c03dd9d64a827aba57775bed0d3

File: Dockerfile (hunk 1 above)
Comment (djmaze) [inline Dockerfile:19]: The current code does not work for me without `flash-attn` installed.  Could you change this to the following? That makes it work:

```Dockerfile
RUN pip install torch packaging wheel &&\
    pip install flash-attn
```

Comment (l4b4r4b4b4) [reply Dockerfile]: hmmm works for me. but I added that as well as all the other requested changes to the PR.

File: Dockerfile (hunk 7 above)
Comment (groovybits) [reply Dockerfile:17]: If not necessary, you may want to remove ffprobe if not used to save space from the static build.

File: Dockerfile (hunk 8 above)
Comment (djmaze) [reply Dockerfile]: I do not see this very often. But I guess many people just don't know. (You can look at the images using `docker history` and see the difference.)

File: Dockerfile [hunk 11]
```diff
@@ -26,8 +26,7 @@ RUN pip install -e .
 RUN apt-get remove -y build-essential && \  
     apt-get autoremove -y && \  
     apt-get clean  && \  
-    rm -rf /var/lib/apt/lists/*  
-
-RUN pip cache purge
+    rm -rf /var/lib/apt/lists/*  && \
+    pip cache purge
```
Comment (djmaze) [inline Dockerfile:31]: This won't make a difference here (nothing installed via pip before.)

File: docker-compose.yml [hunk 12]
```diff
@@ -43,3 +43,29 @@ services:
...
+    networks:
+      - metavoice-net
+    environment:
+      - HF_HOME=/.hf-cache
+    logging:
+      options:
+        max-size: "100m"
+        max-file: "10"
+    healthcheck:
+      test: ["CMD", "curl", "-f", "http://localhost:7861/"]
```
Comment (djmaze) [inline docker-compose.yml:67]: This should be `7860`

**Comment by sidroopdaska** [comment]
> I do not understand why this API is using a very non-standard "X-Payload" header to contain the request body

Will fix this shortly

**Comment by coder543** [comment]
@sidroopdaska With a good enough synthesized voice, I would enjoy being able to paste in an article and have it read it to me sometimes. So, I was just playing around with that kind of thing.

**Review by sidroopdaska — APPROVED** [review]

**Pull request merged**